draft-ietf-jsonpath-base.md

```diff
@@ -902,11 +977,175 @@ size doesn't
+                             index-wild-selector  /  ; ..[*]
+                             "*"                     ; ..*
+                           )
+~~~~
+
+#### Semantics
+{: unnumbered}
+
+The `descendant-selector` is inspired by ECMAScript for XML (E4X). It
+It selects the node and all its descendants..
+
+### Union Selector
+
+#### Syntax
+
+The union selector is syntactically related to the `index-selector`. It contains multiple, comma separated entries.
+
+~~~~ abnf
+union-selector = "[" ws union-entry 1*(ws "," ws union-entry) ws "]"
+
+union-entry    =  ( quoted-member-name /
+                    element-index      /
+                    slice-index
+                  )
+~~~~
+
+> Task (T1): This, besides slice-index, is currently one of only two
+> places in the document that mentions whitespace.  Whitespace needs
+> to be handled throughout the ABNF syntax.  Room Consensus at the
+> 2021-06-15 interim was that JSONPath generally is generous with
+> allowing insignificant whitespace throughout.  Minimizing the impact
+> of the many whitespace insertion points by choosing a rule name such
+> as "S" was mentioned.  Some conventions will probably help with
+> minimizing the number of places where S needs to be inserted.
+
+#### Semantics
+
+A union selects any node which is selected by at least one of the union selectors and selects the concatenation of the
+lists (in the order of the selectors) of nodes selected by the union elements.
+Note that any node selected in more than one of the union selectors is kept
+as many times in the node list.
+
+
+### Filter Selector
+
+#### Syntax
+
+The filter selector has the form `[?<expr>]`. It works via iterating over structured values, i.e. arrays and objects.
+
+~~~~ abnf
+filter-selector    = "[?" boolean-expr "]"
+~~~~
+
+During iteration process each array element or object member is visited and its value — accessible via symbol `@` — or one of its descendants — uniquely defined by a relative path — is tested against a boolean expression `boolean-expr`.
+
+The current item is selected if and only if the result is `true`.
+
+
+~~~~ abnf
+
 
+boolean-expr = logical-expr
+logical-expr = ([neg-op] primary-expr) / logical-or-expr
+neg-op       = "!"                                  ; not operator
+primary-expr = "(" logical-or-expr ")"
+logical-or-expr = logical-and-expr *["||" logical-and-expr]
+logical-and-expr = comp-expr *["&&" comp-expr]
+
+comp-expr    = (rel-path-val) [(comp-op comparable /  ; comparison
+                            regex-op regex     /  ; RegEx test
+                            in-op container )]    ; containment test
+comp-op      = "==" / "!=" /                        ; comparison ...
+               "<"  / ">"  /                        ; operators
+               "<=" / ">="
+regex-op     = "~="                                 ; RegEx match
+in-op        = " in "                               ; in operator
```

Comment:
For the next PR:

Thinking a bit more about `in-op` leads me to the conclusion, that this operator makes sense only for arrays.
`'key' in $.obj` is equivalent to `$.obj.key` and `'key' in <literal-object>` seems to be useless. I see no other rational here. Someone else ?

Reply:
there is still the "enum" use case, as in `$.service_level in ["silver", "gold"]`, which is an abbreviation for `$.service_level == "silver" || $.service_level == "gold"`.

Reply:
well ... this is the `in <array>` use case.